rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+
+- In order to avoid nearly doubling the amount of backend writes for existing large
+clusters (all of whose instances are predominantly ssh services), the existing "node"
+resource (which would be more accurately described as the `ssh_server` resource), will
+be repurposed to represent a server installation which may or may not be running an ssh
+service.  Whether or not other services would also benefit from unification in this way
+can be evaluated on a case-by-case basis down the road.
+
+- In order to support having a single control stream per teleport instance (rather than
+separate control streams for each service) we will need to refactor how instance
+certs are provisioned. Currently, separate certs are granted for each service running
+on a instance, with no single certificate ever encoding all the permissions granted by the
+instance's join token.
+
+Hypothetical GRPC spec:
+
+```protobuf
+// InventoryService is a subset of the AuthService (broken out for the readability)
+service InventoryService {
+    // InventoryControlStream is a bidirectional stream that handles presence and
+    // control messages for peripheral teleport installations.
+    rpc InventoryControlStream(stream ClientMessage) returns (stream ServerMessage);
+}
+
+
+// ClientMessage is sent from the client to the server.
+message ClientMessage {
+    oneof Msg {
+        // Hello is always the first message sent.
+        ClientHello Hello = 1;
+        // Heartbeat periodically updates status.
+        Heartbeat Heartbeat = 2;
+        // LocalScriptInstallResult notifies of installation failures.
+        LocalScriptInstallResult LocalScriptInstallResult = 3;
+    }
+}
+
+// ServerMessage is sent from the server to the client.
+message ServerMessage {
+    oneof Msg {
+        // Hello is always the first message sent.
+        ServerHello Hello = 1;
+        // LocalScriptInstall instructs the client to perform a local-script
+        // upgrade operation.
+        LocalScriptInstall LocalScriptInstall = 2;
+    }
+}
+
+// ClientHello is the first message sent by the client and contains
+// information about the client's version, identity, and claimed capabilities.
+// The client's certificate is used to validate that it has *at least* the capabilities
+// claimed by its hello message. Subsequent messages are evaluated by the limits
+// claimed here.
+message ClientHello {
+    // Version is the currently running teleport version.
+    string Version = 1;
+    // ServerID is the unique ID of the server.
+    string ServerID = 2;
+    // Installers is a list of supported installers (e.g. `local-script`).
+    repeated string Installers = 3;
+
+    // ServerRoles is a list of teleport server roles (e.g. ``).
+    repeated string ServerRoles = 4; 
+}
+
+// Heatbeat periodically 
+message Heartbeat {
+    // TODO
+}
+
+// ServerHello is the first message sent by the server. 
+message ServerHello {
```

Comment:
It kind of does right now, but that's basically a coincidence.  I wanted the key information to all be part of the stream, since its possible that subsequent calls hit different auth servers.  Also, this message is over an authenticated connection, so while it does not currently contain any non-public info, it is able to include it in the future.